File: UserAPI/nodejs_mongodb/controllers/UserController.js
const User = require('../models/user')
const axios = require('axios')

const url = "https://jsonplaceholder.typicode.com/users"

const loadUsers = (req, res) => {
    axios.get(url)
    .then(result => {
        result.data.forEach(user => {
            const User = new User(user)
            User.save()
        })
        res.send("Loading Successful!!")
    })
    .catch(err => {
        console.log(err)
        res.send("error")
    })
}

const returnUsers = (req, res) => {
    User.find({}, (err, usersList) => {
        if(err){
            console.log(err)
            res.status(400).send("error")
        }
        res.status(200).json(usersList)
    })
    .catch(err => {
        console.log(err)
        res.status(400).send("error")
    })
}

const deleteUsers = (req, res) => {
    User.deleteMany({}, err => {
        if(err){
            console.log(err)
            res.status(400).send("error")
        }
        res.status(200).send("Deleted all users")
    })
    .catch(err => {
        console.log(err)
        res.status(400).send("error")
    })
}

const addUser = (req, res) => {
    const user = new User(req.body)
    user.save()
        .then( () => {
            res.status(200).send("User added successfully")
        })
        .catch(err => {
            console.log(err)
            res.status(400).send("error")
        })
}

const getUser = (req, res) => {
    User.findOne({id: req.params.id})
        .then(user => {
            res.status(200).json(user)
        })
        .catch(err => {
            console.log(err)
            res.status(400).send("error")
        })
}

const deleteUser = (req, res) => {
    User.deleteOne({id: req.params.id})
        .then(() => {
            res.status(200).send("Deleted user successfully")
        })
        .catch(err => {
            console.log(err)
            res.status(400).send("Error in deleting")
        })
}

module.exports = {
    loadUsers,
    returnUsers,
    deleteUsers,
    addUser,
    getUser,
    deleteUser
}

File: UserAPI/go_/controllers/userController.go
package controllers

import (
	"encoding/json"
	"io/ioutil"
	"net/http"

	"example.com/user/db"
	"example.com/user/models"
	"github.com/gofiber/fiber/v2"
	"github.com/jinzhu/gorm"
)

func LoadUsers(c *fiber.Ctx) error {
	resp, err := http.Get("https://jsonplaceholder.typicode.com/users")
	if err != nil {
		return c.Status(500).SendString(err.Error())
	}

	respData, err := ioutil.ReadAll(resp.Body)
	if err != nil {
		c.Status(500).SendString(err.Error())
	}

	var userList []models.User
	json.Unmarshal(respData, &userList)

	dbConn := db.DBConn
	for _, user := range userList {
		if err := dbConn.Create(&user).Error; err != nil {
			return c.Status(500).SendString(err.Error())
		}
	}

	return c.Status(201).SendString("Loaded all users to db")
}

func GetUsers(c *fiber.Ctx) error {
	dbConn := db.DBConn
	var users []models.User
	if err := dbConn.Find(&users).Error; err != nil {
		return c.Status(500).SendString(err.Error())
	}
	return c.JSON(users)
}

/* `FIND` queries by ID of gorm and not by the id specified in struct */

func GetUser(c *fiber.Ctx) error {
	id := c.Params("id")
	dbConn := db.DBConn
	var user models.User
	err := dbConn.Find(&user, id).Error
	if err == gorm.ErrRecordNotFound {
		return c.Status(404).SendString("User not found")
	}
	if err != nil {
		return c.Status(500).SendString(err.Error())
	}
	return c.JSON(user)
}

func DeleteUser(c *fiber.Ctx) error {
	id := c.Params("id")
	dbConn := db.DBConn
	var user models.User

	res := dbConn.Find(&user, id)
	if res.Error == gorm.ErrRecordNotFound {
		return c.Status(404).SendString("User not found")
	}
	if res.Error != nil {
		return c.Status(500).SendString(res.Error.Error())
	}

	if err := dbConn.Delete(&user).Error; err != nil {
		return c.Status(500).SendString(err.Error())
	}
	return c.SendStatus(204)
}

func DeleteUsers(c *fiber.Ctx) error {
	dbConn := db.DBConn
	var users []models.User
	if err := dbConn.Delete(&users).Error; err != nil {
		return c.Status(500).SendString(err.Error())
	}
	return c.SendStatus(204)
}

func AddUser(c *fiber.Ctx) error {
	dbConn := db.DBConn
	user := new(models.User)

	if err := c.BodyParser(user); err != nil {
		return c.Status(400).SendString(err.Error())
	}
	if err := dbConn.Create(&user).Error; err != nil {
		return c.Status(500).SendString(err.Error())
	}
	return c.Status(201).SendString("User added to db")
}
